### ISSUE TITLE
Debugging improvements

### ISSUE THREAD

**Issue by lukedays** [cunnane/xloil#7]
### Discussed in https://github.com/cunnane/xloil/discussions/5

Opening an issue just to be more organized!

Ok, so I tried a lot of things and found 2 main problems for VSCode (should be similar for VS 19/22):
- In the launch settings, the `localRoot `and `remoteRoot `folders needed to both point to `${workspaceFolder}` (or else the breakpoints in the UI won't be found)
- It is needed to add to each function `debugpy.debug_this_thread()`. I tried to embed this in a decorator (debug_func in the example below), but with no success. But shouldn't be difficult to add inside the original xloil.func code.

Example: https://github.com/lukedays/xloil-samples/tree/main/hot_reload
- Clone
- Point xlOil.ini to `watcher` module
- Open Excel and add =Watcher()
- Open folder in VSCode and connect
- Changes and breakpoints should work

I was thinking, could this mechanic be packaged into something like `xloil serve`?

**Comment by lukedays** [comment]
I'm trying to transform this mechanic into a script started outside of Excel. The idea is:
- Do `xloil serve`
- Current folder is watched for file changes
- An Excel instance is launched and the functions are loaded
- VSCode connects the debugger to this instance

I took inspiration from xlOil's test runner, but could not make it work. The script exits with no exception when trying to register the `xlo.func` function.
Maybe connecting in some way to the embedded Python to inject the reloader?

`watcher.py`
```python
import importlib
import time
from pathlib import Path

import debugpy
import xloil as xlo
from watchdog.events import FileSystemEventHandler
from watchdog.observers import Observer

debugpy.listen(("localhost", 5678))

app = xlo.Application()
app.visible = True

wb = app.workbooks.add()


def reload_module(path):
    filename = Path(path).name

    if not filename.endswith(".py") or filename in Path(__file__).name:
        return

    filename = filename.replace(".py", "")

    print(f"Trying to load {filename}")
    try:
        m = importlib.import_module(filename)
        importlib.reload(m)
        print(f"Loaded {filename}")
    except Exception as e:
        print(f"Error: {str(e)} on {filename}")


class Handler(FileSystemEventHandler):
    def on_any_event(self, event):
        reload_module(event.src_path)


def watcher():
    current_path = Path(__file__).parent

    # Initial load
    p = current_path.iterdir()
    for f in p:
        if f.is_file():
            reload_module(f)

    # Watch file changes
    obs = Observer()
    obs.schedule(Handler(), path=current_path, recursive=False)
    obs.start()

    # Watch loop
    try:
        while True:
            time.sleep(1)
    finally:
        obs.stop()
        obs.join()


if __name__ == "__main__":
    watcher()

```

`functions.py`

```python
import xloil as xlo

@xlo.func
def Testing():
    return 1
```

**Comment by cunnane** [comment]
This approach is going to be challenging because starting Excel via COM with `app = xlo.Application()` opens a new process, without any loaded addins.  You can then load xlOil with `app.RegisterXLL` but since it's in a different process there is currently no way to interact with it: you'd need a remote server or COM-like interface.  This is possible but it sounds difficult: in some way that is what you want to achieve with the debugger.

I'm going to look at the feasibility of calling `debugpy.debug_this_thread()` everywhere.  If it's just for registered non-async functions, it should be possible to patch `xloil.func` to call this, or just add another decorator before the function.  In full generality though there are quite a few threads which can call your python code:  Excel's calculation threads, RTD server worker threads (used for async), xlOil_Python's background loader and GUI threads.  All of these would need the debugpy treatment.

**Comment by lukedays** [comment]
Alright, I'll do it in the previous way for now. I was able to to hot reloading for deep imports (detect file changes for nested folders/modules), so the iterating process is now way better.

For `debugpy.debug_this_thread()` to not crash when there's no server listening, I had to add:

```python
if debugpy.is_client_connected():
    debugpy.debug_this_thread()
```

**Comment by cunnane** [comment]
Closing this as _debugpy_ support was added in v0.15.  There are some limitations mentioned in [the docs](https://xloil.readthedocs.io/en/stable/xlOil_Python/Debugging.html), in particular VS Code doesn't hit the breakpoints set in RTD functions - it's not clear why as we we are calling `debug_this_thread`.  VS Pro (or Community) is able to debug these functions.

I found `debug_this_thread` won't hang (it does timeout) provided you have already called `listen`, it doesn't need to have a client connected.
